hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/viewfs/ViewFileSystem.java

```diff
@@ -1258,40 +1260,52 @@ public FileStatus getFileStatus(Path f) throws IOException {
             FileStatus status =
                 ((ChRootedFileSystem)link.getTargetFileSystem())
                 .getMyFs().getFileStatus(new Path(linkedPath));
-            result[i++] = new FileStatus(status.getLen(), status.isDirectory(),
-                status.getReplication(), status.getBlockSize(),
-                status.getModificationTime(), status.getAccessTime(),
-                status.getPermission(), status.getOwner(), status.getGroup(),
-                null, path);
+            linkStatuses.add(
+                new FileStatus(status.getLen(), status.isDirectory(),
+                    status.getReplication(), status.getBlockSize(),
+                    status.getModificationTime(), status.getAccessTime(),
+                    status.getPermission(), status.getOwner(),
+                    status.getGroup(), null, path));
           } catch (FileNotFoundException ex) {
             LOG.warn("Cannot get one of the children's(" + path
                 + ")  target path(" + link.getTargetFileSystem().getUri()
                 + ") file status.", ex);
             throw ex;
           }
         } else {
-          result[i++] =
+          internalDirStatuses.add(
               new FileStatus(0, true, 0, 0, creationTime, creationTime,
                   PERMISSION_555, ugi.getShortUserName(),
-                  ugi.getPrimaryGroupName(), path);
+                  ugi.getPrimaryGroupName(), path));
         }
       }
+      FileStatus[] internalDirStatusesMergedWithFallBack = internalDirStatuses
+          .toArray(new FileStatus[internalDirStatuses.size()]);
       if (fallbackStatuses.length > 0) {
-        return consolidateFileStatuses(fallbackStatuses, result);
+        internalDirStatusesMergedWithFallBack =
+            merge(fallbackStatuses, internalDirStatusesMergedWithFallBack);
+      }
+
+      // we don't use target file status as we show the mount link as symlink.
+      if (showMountLinksAsSymlinks) {
+        return merge(internalDirStatusesMergedWithFallBack,
+            linkStatuses.toArray(new FileStatus[linkStatuses.size()]));
```

Comment:
Thanx @umamaheswararao for initiating this, The idea seems good. This fallback mechanism is there in RBF too, through default namespace implementation, shall be great to bring it to ViewFs as well. 
I got the overall idea, Just a doubt regarding the above line.
Why in case `showMountLinkAsSymlinks` is true, we need to give preference to the fallback paths?
Taking example from the structure in `testLSOnLinkParentWithFallbackLinkWithSameMountDirectoryTree`
if we do ls on `/user1/hive/warehouse` the `linkStatuses` will have one partition0 which is mounted and the `internalDirStatusesMergedWithFallBack` will have another partition0 from the fallback. Shouldn't we give preference to the one from the mount entry. Since if I perform an operation on `user1/hive/warehouse/partiton0 the operation will be performed on the mount link resolved path rather than the Fallback partion0?

Here collision is possible only between `linkStatus` and `fallbackStatus`, `linkStatus` and `internalDirStatuses` can not collide since ViewFs isn't allowing to mount /user1/hive -> (some path) as well as /user1/hive/warehouse -> (some path) which is allowed in RBF. So, In case if the exact mount entry is present any call will go to mount entry resolved path and the fallback dir will never be used, so in listing also it could have been overwritten?
Can you help with logic here.

Reply:
Hi @ayushtkn , thank you for questions.

```
>>Just a doubt regarding the above line.
Why in case showMountLinkAsSymlinks is true, we need to give preference to the fallback paths?
```
If you notice in tests, I have set showMountLinkAsSymlinks to false to represent more ViewFSOerloadScheme scenarios where we have showMountLinkAsSymlinks=false bydefault.
In the case showMountLinkAsSymlinks=true, mount link wants to represent as symlink. Here in symlink we will not actually taking any values from targetDirs resolved. This is ViewFileSystem default behavior. In the case of InternalDir we never represent as symlinks. So, this config will not be applicable to internal dir.
Just to be clear: if mount path is /user/hibe/warehouse/partition-0 --> o3fs://....
Here (/user or /user/hive or /user/hive/warehouse ) will be represented as "InternalDirs"
that last dir partition-0 will only be represented as link. So, that configuration will comeinto picture only for partition-0.

In InternalDIr case we are trying get from fallback if they are matching. The reason is, we may want to implement to have mkdirs or other functions to work on internalDir matching path if there is fallback. Example: mkdirs /user/hive/warehouse2
Currently it will not work because we will say /user/hive is readOnly. But the plan is to support and create  /user/hive/warehouse2 in fallback fs. So, it make sense to get the permission/others from fallback dir and represent. Does this make sense? Basically we will want to apply 

>>internalDirStatuses can not collide since ViewFs isn't allowing to mount /user1/hive -> (some path) as well as /user1/hive/warehouse -> (some path) which is allowed in RBF.

Yes, you are right. We don't allow to configure any other mount path in the same tree. Here the idea is if there is fallback, that can be applied everywhere where there is no link configured. That way we can bring the scenario that a base cluster can be a fallback cluster and mount some of the paths from it to some remote fs. So, only links will go to target fs and rest of folder should be pointed to fallback fs. Because fallbackFS is just a single global config, but not specific to any path.

Reply:
>Currently it will not work because we will say /user/hive is readOnly. But the plan is to support and create /user/hive/warehouse2 in fallback fs. So, it make sense to get the permission/others from fallback dir and represent. Does this make sense? 

In most cases it work, For one case I still have a doubt.
But say taking the same tree,  /user1/hive/warehouse/partition-0 is mounted. and the same partion-0 is available through fallback as well.
if I perform any operation say mkdir(/user1/hive/warehouse/partition-0/file) this will go to mount point resolved path not to partion-0 of the fallback path. The above line while listing will show permissions of partion-0 from fall back link, when listing /user1/hive/warehouse, Correct? But actually the mount entry will always be used. The question is which case partion-0 of the fallback will be used above the partion-0 from the mount entry. Even when partion-0 has target path available.

if say in the fallback partion-0 isn't a directory, it is a file, we will show it as a file on listing /user1/hive/warehouse/, but mkdir on (/user1/hive/warehouse/partition-0/file) will be success, Wouldn't it be a inconsistent behavior ? 

I modified your test for my case : 
```
 @Test
  public void testLSOnLinkParentWithFallbackLinkWithSameMountDirectoryTree()
      throws Exception {
    Configuration conf = new Configuration();
    conf.setBoolean(Constants.CONFIG_VIEWFS_MOUNT_LINKS_AS_SYMLINKS, true);
    ConfigUtil.addLink(conf, "/user1/hive/warehouse/partition-0",
        new Path(targetTestRoot.toString()).toUri());
    // Creating multiple directories path under the fallback directory.
    // "/user1/hive/warehouse/partition-0" directory already exists as
    // configured mount point.
    Path file1 = new Path(targetTestRoot,
        "fallbackDir/user1/hive/warehouse/partition-0"); // partition-0 in
    // fallback is a file  ---> Changed this...
    Path dir2 = new Path(targetTestRoot, "fallbackDir/user1/hive/warehouse1");
    fsTarget.create(file1).close();
    fsTarget.mkdirs(dir2);
    URI viewFsUri = new URI(FsConstants.VIEWFS_SCHEME,
        Constants.CONFIG_VIEWFS_DEFAULT_MOUNT_TABLE, "/", null, null);

    ConfigUtil
        .addLinkFallback(conf, new Path(targetTestRoot, "fallbackDir").toUri());

    try (FileSystem vfs = FileSystem.get(viewFsUri, conf)) {
      for (FileStatus stat : vfs.listStatus(
          new Path(viewFsUri.toString(), "/user1/hive/warehouse/"))) {
        if (file1.getName().equals(stat.getPath().getName())) {
          assertTrue(stat.isFile()); // listing says
          // /user1/hive/warehouse/partition-0 is a file.
          Path fileUnderDir = new Path(stat.getPath(), "check");
          assertTrue(vfs.mkdirs(fileUnderDir)); // Created a file under file?

        }
      }
    }
  }
```
Is this behavior fine?

Reply:
HI @ayushtkn ,

The file actually creates under the above mount link you added.
```
ConfigUtil.addLink(conf, "/user1/hive/warehouse/partition-0",
        new Path(targetTestRoot.toString()).toUri());
```
Once you have mounted a path, links are more prioritized in ViewFS. I don't see a requirement people will configure files as mount links. We only support dirs. That dir takes the precedence. You will loose complete access to fallback, once you added mount path link. Tricky case here is, when you have matching file at fallback, you no more see that file. That you configured to represent as dir from mount view. we don't create anything under file in fallback fs. 

in SymLinks enabled, isFile always should be false. I think raised a good point that we don't need that if condition where its making fallback high precedence and making think wrong in symlink enabled case. I will removed it and verify. I think I overthought there.
To make thing clear:

Mount tree links will have higher precendence.
Fallback dirs have higher precedence compared to moount tree internal dir.
If no matching mount tree internal dirs will have as last precedence.